namer/zk-leader/src/main/scala/io/buoyant/namer/zkLeader/ZkLeaderNamer.scala

```diff
@@ -0,0 +1,85 @@
+package io.buoyant.namer.zkLeader
+
+import com.twitter.common.zookeeper.Group.GroupChangeListener
+import com.twitter.common.zookeeper._
+import com.twitter.finagle.util.InetSocketAddressUtil
+import com.twitter.finagle.{Group => _, _}
+import com.twitter.util._
+import java.net.InetSocketAddress
+import org.apache.zookeeper.KeeperException.NoNodeException
+import org.apache.zookeeper.data.ACL
+import scala.collection.JavaConverters._
+
+/**
+ * This namer accepts paths of the form /<prefix>/<zkPath>.  The zkPath is the location
+ * in ZooKeeper of a leader group.  This namer resolves to the addresses stored in the data of
+ * the leader of the group.
+ */
+class ZkLeaderNamer(
+  prefix: Path,
+  hosts: String,
+  factory: Iterable[InetSocketAddress] => ZooKeeperClient
+) extends Namer {
+
+  def this(prefix: Path, hosts: String) = this(
+    prefix,
+    hosts,
+    h => new ZooKeeperClient(ZooKeeperUtils.DEFAULT_ZK_SESSION_TIMEOUT, h.asJava)
+  )
+
+  override def lookup(path: Path): Activity[NameTree[Name]] = bind(path)
+
+  private[this] val client = factory(InetSocketAddressUtil.parseHosts(hosts).toSet)
+
+  private[this] def bind(path: Path, residual: Path = Path.empty): Activity[NameTree[Name]] = {
+    val id = prefix ++ path
+
+    val group = new Group(client, Iterable.empty[ACL].asJava, path.show)
+    val candidate = new CandidateImpl(group)
+    leaderAddr(candidate).map { initial =>
+      Var.async[Addr](initial) { update =>
+        val stop = group.watch(
+          new GroupChangeListener {
+            override def onGroupChange(memberIds: java.lang.Iterable[String]): Unit = {
+              leaderAddr(candidate) match {
+                case Some(addr) => update() = addr
+                case None => update() = Addr.Neg
+              }
+            }
+          }
+        )
+
+        Closable.make { _ =>
+          stop.execute()
+          Future.Unit
+        }
+      }
+    } match {
+      case Some(addr) =>
+        val tree = addr.map[Activity.State[NameTree[Name]]] {
+          case bound: Addr.Bound => Activity.Ok(NameTree.Leaf(Name.Bound(addr, id, residual)))
+          case _ => Activity.Ok(NameTree.Neg)
+        }
+        Activity(tree)
+      case None =>
+        if (!path.isEmpty) {
+          val n = path.size
+          bind(path.take(n - 1), residual ++ path.drop(n - 1))
+        } else {
+          Activity.value(NameTree.Neg)
+        }
+    }
+  }
+
+  private[this] def leaderAddr(candidate: Candidate): Option[Addr] = {
+    try {
+      Option(candidate.getLeaderData.orNull).map { data =>
```

Comment:
getLeaderData returns a `com.google.common.base.Optional`.  This is this simplest way I could think of to convert it to a `scala.Option`